### ISSUE TITLE
Failed to remove output annotation in Quantizer for PT2 QAT quantization

### ISSUE THREAD

**Issue by leslie-fang-intel** [pytorch/pytorch#112833]
### 🐛 Describe the bug

As in https://github.com/pytorch/pytorch/pull/112140, we plan to remove the output annotation in `X86InductorQuantizer`, it works for PTQ Quantization, but failed for QAT case. Try to debug this issue and find 2 issues:

- Issue 1: Current `_get_input_output_quantized_filter` implementation, will filter out any pattern without output quant. So, for a QAT `Conv-BN` pattern, if we didn't annotate its output explicitly or connect to any pattern with input annotated, it will failed to match.
https://github.com/pytorch/pytorch/blob/53fff56ab8fead5b8829367730363c13e206975b/torch/ao/quantization/pt2e/qat_utils.py#L183-L193
- Issue 2:  For a  pattern
```
Conv1       Conv2
    \          /
    BN1    BN2
      \     /
        Add
```
We will annotate the extra input of add (which is connected to BN2 in this case) to perform `conv->add` fusion https://github.com/pytorch/pytorch/blob/94ebf52ea326dec22300a2d7bacd8a15d8f0c12f/torch/ao/quantization/quantizer/x86_inductor_quantizer.py#L474-L480
However, `Conv2->BN2` will be replaced to a new `Conv->BN` pattern for QAT training in `_fuse_conv_bn_qat`. So, we may need some update node name of `Add`'s `quantization_annotation`, similar as `_update_conv_input_qspec_map_after_replacement` here https://github.com/pytorch/pytorch/blob/53fff56ab8fead5b8829367730363c13e206975b/torch/ao/quantization/pt2e/qat_utils.py#L661-L662

### Versions

Collecting environment information...
PyTorch version: 2.2.0a0+gitca59581
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 11.2.1 20210728 (Red Hat 11.2.1-1)
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.17

Python version: 3.9.10 (main, Mar  2 2022, 12:02:00)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-4.19.5-1.el7.elrepo.x86_64-x86_64-with-glibc2.17
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                56
On-line CPU(s) list:   0-55
Thread(s) per core:    1
Core(s) per socket:    28
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel Genuine CPU
Stepping:              10
CPU MHz:               1878.393
CPU max MHz:           2900.0000
CPU min MHz:           1200.0000
BogoMIPS:              5800.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              39424K
NUMA node0 CPU(s):     0-27
NUMA node1 CPU(s):     28-55
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req pku ospke avx512_vnni flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-coding==1.3.3
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] intel-extension-for-pytorch==2.2.0+git2341961
[pip3] mypy==1.4.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.3
[pip3] torch==2.2.0a0+gitfd2203a
[pip3] torchvision==0.16.0a0+370134
[pip3] triton==2.0.0
[conda] mkl                       2023.0.0            intel_25398    intel
[conda] mkl-include               2023.0.0                  <pip>
[conda] mkl-include               2023.0.0            intel_25398    intel
[conda] mkl-service               2.4.0           py39h3605609_14    intel
[conda] mkl-static                2023.0.0                  <pip>
[conda] mkl_fft                   1.3.1           py39hcab1719_22    intel
[conda] mkl_random                1.2.2           py39hbf47bc3_22    intel
[conda] mkl_umath                 0.1.1           py39hf66a691_32    intel
[conda] numpy                     1.22.3           py39hf0956d0_5    intel
[conda] numpy                     1.24.3                    <pip>
[conda] numpy-base                1.22.3           py39h45c9ace_5    intel
[conda] torchvision               0.16.0a0+370134           <pip>
[conda] triton                    2.0.0                     <pip>

**Comment by leslie-fang-intel** [comment]
Hi @andrewor14 @jerryzh168, Could you kindly help to take a look of this issue? 
Here is some of my thought:
- For Issue 1, Is it possible that we remove the hard limitation that the pattern must have a output quant?
- For Issue 2, How about we add a function to update `Add`'s `quantization_annotation` just as `_update_conv_input_qspec_map_after_replacement` https://github.com/pytorch/pytorch/blob/53fff56ab8fead5b8829367730363c13e206975b/torch/ao/quantization/pt2e/qat_utils.py#L661-L662

**Comment by andrewor14** [comment]
Hi @leslie-fang-intel, just wanted to understand issue (2) better. This is not specific to "add" right? Any node following the `conv-bn` pattern that refers to nodes replaced in that pattern will have the same problem. Is that true? If so I feel we may want a more general solution.

For this specific case, can you insert the observer as an output of the batchnorm getitem node, instead of as the input edge between batchnorm getitem and the add nodes?

**Comment by leslie-fang-intel** [comment]
Hi @andrewor14, thanks for the feedback.

> This is not specific to "add" right? Any node following the conv-bn pattern that refers to nodes replaced in that pattern will have the same problem. Is that true? If so I feel we may want a more general solution.

Yean, I think so. If we annotate a node's input edge connected with a `conv-bn` pattern, and the `get-item` node in `bn` being replaced in QAT flow will have this problem. 

> For this specific case, can you insert the observer as an output of the batchnorm getitem node, instead of as the input edge between batchnorm getitem and the add nodes?

This is exactly same as current behavior. However, we are modifying current recipe to only annotate the input nodes of conv-bn as  in https://github.com/pytorch/pytorch/pull/112140 for PTQ. The reason we only annotate the inputs, because we want conv/GEMM pattern outputs a Float32 tensor if there is no q-dq afterwards.
